src/wp-includes/block-template-utils.php

```diff
@@ -1,23 +1,481 @@
+ * @since 5.9.0
+ *
+ * @param string $type Template part area name.
+ *
+ * @return string Input if supported, else the uncategorized value.
+ */
+function _filter_block_template_part_area( $type ) {
+	$allowed_areas = array_map(
+		function ( $item ) {
+			return $item['area'];
+		},
+		get_allowed_block_template_part_areas()
+	);
+	if ( in_array( $type, $allowed_areas, true ) ) {
+		return $type;
+	}
+
+	/* translators: %1$s: Template area type, %2$s: the uncategorized template area value. */
+	$warning_message = sprintf( __( '"%1$s" is not a supported wp_template_part area value and has been added as "%2$s".' ), $type, WP_TEMPLATE_PART_AREA_UNCATEGORIZED );
+	trigger_error( $warning_message, E_USER_NOTICE );
+	return WP_TEMPLATE_PART_AREA_UNCATEGORIZED;
+}
+
+/**
+ * Finds all nested template part file paths in a theme's directory.
+ *
+ * @access private
+ * @since 5.9.0
+ *
+ * @param string $base_directory The theme's file path.
+ * @return array $path_list A list of paths to all template part files.
+ */
+function _get_block_templates_paths( $base_directory ) {
+	$path_list = array();
+	if ( file_exists( $base_directory ) ) {
+		$nested_files      = new RecursiveIteratorIterator( new RecursiveDirectoryIterator( $base_directory ) );
+		$nested_html_files = new RegexIterator( $nested_files, '/^.+\.html$/i', RecursiveRegexIterator::GET_MATCH );
+		foreach ( $nested_html_files as $path => $file ) {
+			$path_list[] = $path;
+		}
+	}
+	return $path_list;
+}
+
+/**
+ * Retrieves the template file from the theme for a given slug.
+ *
+ * @access private
+ * @since 5.9.0
+ *
+ * @param string $template_type wp_template or wp_template_part.
+ * @param string $slug template slug.
+ *
+ * @return array|null Template.
+ */
+function _get_block_template_file( $template_type, $slug ) {
+	$template_base_paths = array(
+		'wp_template'      => 'block-templates',
+		'wp_template_part' => 'block-template-parts',
+	);
+	$themes              = array(
+		get_stylesheet() => get_stylesheet_directory(),
+		get_template()   => get_template_directory(),
+	);
+	foreach ( $themes as $theme_slug => $theme_dir ) {
+		$file_path = $theme_dir . '/' . $template_base_paths[ $template_type ] . '/' . $slug . '.html';
+		if ( file_exists( $file_path ) ) {
+			$new_template_item = array(
+				'slug'  => $slug,
+				'path'  => $file_path,
+				'theme' => $theme_slug,
+				'type'  => $template_type,
+			);
+
+			if ( 'wp_template_part' === $template_type ) {
+				return _add_block_template_part_area_info( $new_template_item );
+			}
+
+			if ( 'wp_template' === $template_type ) {
+				return _add_block_template_info( $new_template_item );
+			}
+
+			return $new_template_item;
+		}
+	}
+
+	return null;
+}
+
+/**
+ * Retrieves the template files from  the theme.
+ *
+ * @access private
+ * @since 5.9.0
+ *
+ * @param string $template_type wp_template or wp_template_part.
+ *
+ * @return array Template.
+ */
+function _get_block_templates_files( $template_type ) {
```

Comment:
Should this be `_get_block_template_files`? (no plural in `templates`)

Reply:
I hesitated but I choice "templates" because it's multiple templates while `template-files` could mean mutiple files for a single template. Not entirely confident with my choice, that said, it's not very important because it's a private function.